### ISSUE TITLE
No output when input is standard input

### ISSUE THREAD

**Issue by pkgw** [Lindydancer/e2ansi#1]
The e2ansi-cat program seems to produce no useful output when the input data comes from stdin. For instance,

```
emacs --batch -Q -l ~/.emacs -l bin/e2ansi-cat /etc/passwd
```

works fine for me, but

```
cat /etc/passwd |emacs --batch -Q -l ~/.emacs -l bin/e2ansi-cat /dev/stdin
```

produces no output. In situations where you know the input is stdin, you *could* save it to a temporary file, but this is problematic for streaming data and such. Is there a way to make it so that it can handle stdin directly?

**Comment by Lindydancer** [comment]
I left out this for two reasons:

* The file name extension is one of the things Emacs considers when selecting major mode. When reading from stdin, it's not available.

* When executed without any argument, I prefer for the command to print its usage information. In this case, this is especially important since Emacs eats the "--help" option and displays its own help information.

However, I am thinking about using the convention that if a file name "-" is specified, I would read from standard input. Also, I could add a "--mode" command line option, that way a user could specify the major mode that should be used. What do you think about this?

**Comment by pkgw** [comment]
In my example I was providing `/dev/stdin` as a filename argument, so it should have at least produced *some* output. So I suspect that there is a genuine bug somewhere. I would agree that using "-" to mean standard input would be a feature that many people would expect, though I think all modern operating systems have either `/dev/stdin` or `/dev/fd/0` to let you tell a program to process stdin even if it doesn't support "-" as an alias.

I personally would like to use `e2ansi` even for stdin-type cases where Emacs is unable to use the filename to help guess the major mode, if that's possible on the Emacs side.

**Comment by Lindydancer** [comment]
I don't think Emacs understands opening /dev/stdin. However, it's possible to do repeat `read-string' until an error is issues. (There is one minor issue, it can't tell if the last line ended with a newline or not, but I can live with that.)

I will give it a try as soon as I find some free time to do it on.

**Comment by Lindydancer** [comment]
I just published a new version. If you give "-" as file name, it reads stdin. --mode can be used to specify major mode.

**Comment by pkgw** [comment]
Great! I'd say that that fixes this issue.

**Comment by Lindydancer** [comment]
I just discovered that new versions of `less` supports using input preprocessors as filters. If `LESSOPEN` starts with `|-` then the input preprocessor will started with `-` as the file name, and it's expected to read from standard output.

Using `magic-mode-alist`, it's possible to teach Emacs to recognise some file types based on file content. I have provided a (yet small) file `e2ansi-magic.el` that adds support for svn-style diffs -- I expect to fill the file as I find new uses.

In other words, using something like:

    export "LESSOPEN=|-emacs --batch -Q -l ~/.emacs -l e2ansi-magic.el -l bin/e2ansi-cat %s"

It's possible enjoy syntax highlighted output to things like:

    svn diff | less

I have updated the documentation to reflect this. Also, I posted the package on Melpa a few minutes ago, hopefully it will appear soon.

**Comment by pkgw** [comment]
Yes, that's the context in which I'm using `e2ansi` and it's working great!

Note also that even newer versions of `less` have an additional feature that you might be interested in:

    Note that a preprocessor cannot output an empty file, since that is
    interpreted as meaning there is no replacement, and the original file is
    used. To avoid this, if LESSOPEN starts with two vertical bars, the exit
    status of the script becomes meaningful. If the exit status is zero, the
    output is considered to be replacement text, even if it empty. If the exit
    status is nonzero, any output is ignored and the original file is used.
    For compatibility with previous versions of less, if LESSOPEN starts with
    only one vertical bar, the exit status of the preprocessor is ignored.

However, older versions of `less` don't understand the double-bar syntax and will give an error if they see it, and many of these older versions of `less` are still in the wild.

**Comment by Lindydancer** [comment]
Interesting, I didn't know that. I will check whether or not `e2ansi-cat` sets the exit code appropriately, it could be useful in other situations as well.

When it comes to the `||` syntax -- normally the only case when `e2ansi-cat` would output an empty file is when the input file is empty, in which case it doesn't matter if the output or the original file is shown by `less`. However, I could imagine special cases where Emacs would perform transformations, like uncompressing a file. In that case the output could be empty even though `less` should not present the original file. In other words, it might be worth mentioning this in the `e2ansi` documentation.
